src/finite_difference.c

```diff
@@ -0,0 +1,65 @@
+#include "finite_difference.h"
+#include <stdio.h>
+#include <string.h>
+
+/// @brief Approximates the gradient of a function using the finite difference
+/// method
+
+/// @param func The function for which to approximate the gradient
+/// @param point The point at which to appriximate the gradient
+/// @param grad Return parameter containing the approximated gradient
+/// @param n Number of dimensions of the function
+/// @param eps Size of the step taken away from point for the approximation.
+/// Lower eps should lead to better approximations. Should not be smaller than
+/// the machine floating point precision.
+/// @param type Type of finite difference approximation (Forward, Backward or
+/// Central)
+void gradientApproximation(function func, real point[], real grad[], int n,
+                           real eps, approximationType type) {
+  // It is imperative to choose h so that x and x + h differ by an exactly
+  // representable number
+  // see:
+  // http://www.it.uom.gr/teaching/linearalgebra/NumericalRecipiesInC/c5-7.pdf
+  real h_next[n];
+  real h_prev[n];
+  for (int i = 0; i < n; i++) {
+    volatile real temp = point[i] + eps;
```

Comment:
This is something recommended by the book numerical recipes. Here is the paragraph. 

Always choose h so that x+ h and x differ by an exactly
representable number. This can usually be accomplished by the program steps
temp = x + h
h = temp − x
Some optimizing compilers, and some computers whose floating-point chips have
higher internal accuracy than is stored externally, can foil this trick; if so, it is
usually enough to declare temp as volatile,